tls/s2n_config.c

```diff
@@ -562,6 +571,50 @@ static int s2n_config_add_cert_chain_and_key_impl(struct s2n_config *config, str
     return S2N_SUCCESS;
 }
 
+S2N_RESULT s2n_config_validate_certificate_preferences(const struct s2n_config *config,
+        const struct s2n_security_policy *security_policy)
```

Comment:
Definitely agree with the specific error thing, added!

I also agree that it's tricky, complex, and generally unpleasant, and the likelihood of missing something is much higher that I'd like it to be. 🙃 

My idea with this feature is that there should be no scenario in which a customer can both
1. set a security policy successfully
2. use certs that are incompatible with that security policy

I like your last point about a sort of builder mechanism, and agree that it would make this logic much more simple. However, I don't see any (non breaking/backwards compatible) way to force customers to set their security policy before loading certs.

> I wonder if we could do something like make the validation apply specifically to when certs are added to a config after the security policy is configured.

Without any mechanism to enforce setting the security policy before certs, then I don't like this approach because a customer that didn't read our documentation and called the APIs in the wrong order wouldn't know that they were in violation of their security policy.

Reply:
I guess we should assume that nobody is modifying their config after associating it with a connection though?

Reply:
> what happens if a non-RFC9151 security policy is set on a config with no certs loaded, and then the config is associated with a connection that has an RFC9151 security policy.

This operation is expected to fail. This check is handled by the logic on line 287 of `s2n_connection.c`. If there is a security policy override on the connection, then the config will be validated against the override. If that validation is not successful then `s2n_connection_set_config` will fail.

This behavior is also covered in the "set a security_policy_override then set an invalid config" test case in s2n_security_policy_cert_preferences_test.

> Assuming there will be cases like these that can't really be enforced

My goal was to enforce this behavior in _all_ cases, but if I've missed one let me know!

Reply:
Talked offline. I'm missing the last case
1. create default config
2. associate default config with connection
3. set connection security policy to rfc9151
4. load invalid cert into default config

It is going to be generally very painful to make the 4th step fail.

That being said I'm still in favor of trying to validate as many cases as possible. I think the above scenario is quite different than the previous discussed scenarios in that it is much less likely to show up in customer code. In comparison, if we only do security policy validations when the the security policy has already been set, I expect us to miss about 50% of customers.